bisection.go

```diff
@@ -74,63 +67,55 @@ func (b *Bisection) Iterate(f, g float64) (Operation, float64, error) {
 	if b.minF < minF {
 		minF = b.minF
 	}
-	if StrongWolfeConditionsMet(f, g, minF, b.initGrad, b.currStep, 0, b.GradConst) {
-		b.lastOp = MajorIteration
-		return b.lastOp, b.currStep, nil
-	}
-
-	// Deciding on the next step size
-	if math.IsInf(b.maxStep, 1) {
-		// Have not yet bounded the minimum
-		switch {
-		case g > 0:
-			// Found a change in derivative sign, so this is the new maximum
-			b.maxStep = b.currStep
-			b.maxF = f
-			b.maxGrad = g
-			return b.nextStep((b.minStep + b.maxStep) / 2)
-		case f <= b.minF:
-			// Still haven't found an upper bound, but there is not an increase in
-			// function value and the gradient is still negative, so go more in
-			// that direction.
-			b.minStep = b.currStep
-			b.minF = f
-			b.minGrad = g
-			return b.nextStep(b.currStep * 2)
-		default:
-			// Increase in function value, but the gradient is still negative.
-			// Means we must have skipped over a local minimum, so set this point
-			// as the new maximum
-			b.maxStep = b.currStep
-			b.maxF = f
-			b.maxGrad = g
-			return b.nextStep((b.minStep + b.maxStep) / 2)
+	if b.lastOp == FuncEvaluation {
+		// See if the function value is good enough to make progress. If it is,
+		// evaluate the gradient. If not, set it to the upper bound if the bound
+		// has not yet been found, otherwise iterate toward the minimum location.
+		if f <= b.minF {
+			b.lastOp = GradEvaluation
+			return b.lastOp, b.currStep, nil
 		}
-	}
-
-	// Already bounded the minimum, but wolfe conditions not met. Need to step to
-	// find minimum.
-	if f <= b.minF && f <= b.maxF {
-		if g < 0 {
-			b.minStep = b.currStep
-			b.minF = f
-			b.minGrad = g
-		} else {
+		if math.IsInf(b.maxStep, 1) {
 			b.maxStep = b.currStep
 			b.maxF = f
-			b.maxGrad = g
+			return b.nextStep((b.minStep + b.maxStep) / 2)
 		}
-	} else {
-		// We found a higher point. Want to push toward the minimal bound
 		if b.minF <= b.maxF {
 			b.maxStep = b.currStep
 			b.maxF = f
-			b.maxGrad = g
 		} else {
 			b.minStep = b.currStep
 			b.minF = f
-			b.minGrad = g
 		}
+		return b.nextStep((b.minStep + b.maxStep) / 2)
+	}
+	// The function value was lower. Check if this location is sufficient to
+	// converge the linesearch, otherwise iterate.
+	if StrongWolfeConditionsMet(f, g, minF, b.initGrad, b.currStep, 0, b.GradConst) {
```

Comment:
Wouldn't it be better replace this with an explicit check? Only the curvature condition needs to be checked and we wouldn't have to repeatedly call Abs on the initial derivative (not that it is a big cost). We will have to change the detection of convergence anyway to deal with the floating point noise.